### PR TITLE
SB-29645: CephS3 config/support for Flink checkpointing, Druid deep storage, data-products cloud storage and other general fixes

### PR DESCRIPTION
**Pull request by shobhit-vashistha**
CHANGES

ceph: spark-provision
ceph: data-products-deploy
ceph: data-pipeline-flink
seperate kafka clusters fix
druid s3 config
druid config fixes + router config
data-products-deploy fixes
data-pipeline-flink redis increment functonality
data-pipeline-flink output tag conflict fix

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

### Type of change

Please choose appropriate options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### How Has This Been Tested?

Please describe the tests that you ran to verify your changes in the below checkboxes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Ran Test A
- [ ] Ran Test B

**Test Configuration**:
* Software versions:
* Hardware versions:

### Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: ansible/roles/druid-ingestion/defaults/main.yml [hunk 1]
```diff
@@ -1,8 +1,13 @@
 remote_user: druid
+remote_group: users
```
Comment (anandp504) [inline ansible/roles/druid-ingestion/defaults/main.yml:2]: Do we need a remote_group called `users` defined here?

File: ansible/roles/secor-telemetry-backup-deploy/templates/log4j.j2 [hunk 2]
```diff
@@ -11,7 +11,7 @@ log4j.appender.CONSOLE.layout.ConversionPattern=%d{ISO8601} [%t] (%C:%L) %-5p %m
 
 log4j.appender.ROLLINGFILE=org.apache.log4j.DailyRollingFileAppender
 log4j.appender.ROLLINGFILE.Threshold=INFO
-log4j.appender.ROLLINGFILE.File=/mount/{{ secor_service_name }}/logs/secor.log
+log4j.appender.ROLLINGFILE.File=/mount/{{ item.0 }}/logs/secor.log
```
Comment (anandp504) [inline ansible/roles/secor-telemetry-backup-deploy/templates/log4j.j2:14]: The secor-telemetry-backup-deploy uses a dictionary named `secor_service_name` to defined all the required backups. The `{{ item.0 }}` referred here assumes the role will have a dictionary named item. Can you elaborate a little bit more on this change?

File: data-pipeline-flink/dp-core/src/main/scala/org/sunbird/dp/core/cache/DataCache.scala [hunk 3]
```diff
@@ -131,7 +131,7 @@ class DataCache(val config: BaseJobConfig, val redisConnect: RedisConnect, val d
     try {
       set(key, value);
     } catch {
-      case ex@(_: JedisConnectionException | _: JedisException) =>
+      case ex: JedisException =>
```
Comment (anandp504) [inline data-pipeline-flink/dp-core/src/main/scala/org/sunbird/dp/core/cache/DataCache.scala:134]: `JedisConnectionException` class extends from JedisConnection. We might have added it to specifically capture a JedisConnectionException. Is this change necessary?

File: data-pipeline-flink/dp-core/src/main/scala/org/sunbird/dp/core/job/BaseJobConfig.scala [hunk 4]
```diff
@@ -21,6 +21,7 @@ class BaseJobConfig(val config: Config, val jobName: String) extends Serializabl
   val kafkaProducerMaxRequestSize: Int = config.getInt("kafka.producer.max-request-size")
   val kafkaProducerBatchSize: Int = config.getInt("kafka.producer.batch.size")
   val kafkaProducerLingerMs: Int = config.getInt("kafka.producer.linger.ms")
+  val kafkaProducerCompression: String = config.getString("kafka.producer.compression")
```
Comment (anandp504) [inline data-pipeline-flink/dp-core/src/main/scala/org/sunbird/dp/core/job/BaseJobConfig.scala:24]: We will need the default compression to be set to snappy. The cpu utilization with snappy is definitely lower and hence we have defaulted to snappy.

File: data-pipeline-flink/dp-core/src/main/scala/org/sunbird/dp/core/job/BaseJobConfig.scala [hunk 5]
```diff
@@ -59,7 +60,9 @@ class BaseJobConfig(val config: Config, val jobName: String) extends Serializabl
     properties.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaProducerBrokerServers)
     properties.put(ProducerConfig.LINGER_MS_CONFIG, new Integer(kafkaProducerLingerMs))
     properties.put(ProducerConfig.BATCH_SIZE_CONFIG, new Integer(kafkaProducerBatchSize))
-    properties.put(ProducerConfig.COMPRESSION_TYPE_CONFIG, "snappy")
+    if (kafkaProducerCompression != null && kafkaProducerCompression.nonEmpty) {
```
Comment (anandp504) [inline data-pipeline-flink/dp-core/src/main/scala/org/sunbird/dp/core/job/BaseJobConfig.scala:63]: With a default value, we need not do a check on whether kafkaProducerCompression is set.

File: ansible/inventory/env/group_vars/all.yml [hunk 6]
```diff
@@ -170,4 +170,15 @@ processing_zookeeper_ip: "{{ groups['processing-cluster-zookeepers'][0] }}"
 ### docker image pull secrets name and this will be created in kubernetes
 imagepullsecrets: "{{env}}-registry-secret"
 dialcode_host: "http://{{private_ingressgateway_ip}}/dial"
-dialcode_endpoint: "dialcode/v3/read/"
\ No newline at end of file
+dialcode_endpoint: "dialcode/v3/read/"
+
+# s3 storage config
+s3_storage_key: ""
+s3_storage_secret: ""
```
Comment (anandp504) [inline ansible/inventory/env/group_vars/all.yml:177]: The `s3_storage_secret` has to go into ansible secrets as the secret will be encrypted and stored in the ansible vault.

**Review by anandp504 — CHANGES_REQUESTED** [review]

File: ansible/inventory/env/group_vars/all.yml [hunk 7]
```diff
@@ -26,7 +26,7 @@ env_name: "{{env}}"
 kafka_brokers: "{{groups['processing-cluster-kafka']|join(':9092,')}}:9092"
 zookeepers: "{{groups['zookeeper']|join(':2181,')}}:2181"
 # Ingestion cluster
-ingestion_kafka_brokers: "{{groups['processing-cluster-kafka']|join(':9092,')}}:9092"
+ingestion_kafka_brokers: "{{groups['ingestion-cluster-kafka']|join(':9092,')}}:9092"
```
[the remark this answers is not in the document]
Comment (shobhit-vashistha) [reply ansible/inventory/env/group_vars/all.yml:29]: sure, I assumed it was an unintentional miss, will remove from the PR

File: ansible/inventory/env/group_vars/all.yml [hunk 8]
```diff
@@ -170,4 +170,15 @@ processing_zookeeper_ip: "{{ groups['processing-cluster-zookeepers'][0] }}"
...
+
+# s3 storage config
+s3_storage_key: ""
+s3_storage_secret: ""
+s3_storage_endpoint: ""
+s3_request_signature_version: AWS4-HMAC-SHA256
+s3_path_style_access: false
+s3_https_only: false
+s3_default_bucket_location: ""
+s3_storage_container: "igot"
```
[the remark this answers is not in the document]
Comment (shobhit-vashistha) [reply ansible/inventory/env/group_vars/all.yml:183]: s3_default_bucket_location is for the location of the bucket (e.g. us-east-1). For iGOT's use case, NIC wants to be able to configure one bucket for publicly accessible resources like downloadable certificates and other media files, and another bucket for private resources like checkpointing data, deep storage data, secor backups, logs etc.

Will remove the default value `igot` and `igotlogs` from here

File: ansible/roles/analytics-druid/defaults/main.yml [hunk 9]
```diff
@@ -181,6 +206,19 @@ default_druid_configs:
    #Logging the indexing logs to azure
    druid_log_azure_container: "{{ druid_azure_container_name }}"
    druid_log_azure_folder: "druidlogs"
+   #Druid S3 Details
+   druid_storage_type: "{{ druid_storage_type }}"
+   s3_access_key: "{{ s3_storage_key }}"
+   s3_secret_key: "{{ s3_storage_secret }}"
+   s3_bucket: "{{ s3_dp_storage_container }}"
+   s3_endpoint: "{{ s3_storage_endpoint }}"
+   s3_segment_dir: "druid/raw/segments"
```
[the remark this answers is not in the document]
Comment (shobhit-vashistha) [reply ansible/roles/analytics-druid/defaults/main.yml:215]: Got it, will add different values for `s3_segment_dir`, in `raw` and `rollup` section of `default_druid_configs` dictionary

File: ansible/roles/analytics-druid/defaults/main.yml [hunk 10]
```diff
@@ -181,6 +206,19 @@ default_druid_configs:
...
+   s3_path_like_access: {{ s3_path_style_access }}
+   s3_v4_sign_region: "{{ s3_default_bucket_location }}"
+   #Logging the indexing logs to s3
+   s3_logging_bucket: "{{ s3_dp_storage_container }}"
+   s3_indexer_logs_dir: "druid/raw/stage/indexing_logs"
```
[the remark this answers is not in the document]
Comment (shobhit-vashistha) [reply ansible/roles/analytics-druid/defaults/main.yml:220]: Got it, will add different values for `s3_indexer_logs_dir`, in `raw` and `rollup` section of `default_druid_configs` dictionary

File: ansible/roles/analytics-spark-provision/templates/jets3t.j2 [hunk 11]
```diff
@@ -1,3 +1,16 @@
-storage-service.request-signature-version=AWS4-HMAC-SHA256
-s3service.s3-endpoint=s3-ap-south-1.amazonaws.com
+storage-service.request-signature-version={{ jets3t_s3_request_signature_version }}
+s3service.s3-endpoint={{ jets3t_s3_endpoint_host }}
+
+{% if jets3t_s3_disable_dns_buckets %}
+s3service.disable-dns-buckets=true
```
[the remark this answers is not in the document]
Comment (shobhit-vashistha) [reply ansible/roles/analytics-spark-provision/templates/jets3t.j2:5]: spark s3 connector does not hit the custom endpoint until the value for `s3service.s3-endpoint` is configured in `jets3t.j2`. same goes for the other configuration. 

`jets3t_s3_disable_dns_buckets` can be directly rendered instead of the if statements (motivation was, that the `s3service.disable-dns-buckets` config is not needed in case of AWS S3), will update this accordingly

File: ansible/roles/analytics-spark-provision/defaults/main.yml [hunk 12]
```diff
@@ -32,3 +32,12 @@ hadoop_aws_url: https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-aws/2.7.
 java_xmlbuilder_url: https://repo1.maven.org/maven2/com/jamesmurty/utils/java-xmlbuilder/1.1/java-xmlbuilder-1.1.jar
 hadoop_azure_url: https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-azure/2.7.3/hadoop-azure-2.7.3.jar
 azure_storage: https://repo1.maven.org/maven2/com/microsoft/azure/azure-storage/3.0.0/azure-storage-3.0.0.jar
+
+
+# jets3t s3 config, allows us to configure for s3-like object stores
+jets3t_s3_request_signature_version: "{{ s3_request_signature_version }}"
+jets3t_s3_endpoint_host: "{% if s3_storage_endpoint %}{{ s3_storage_endpoint | regex_replace('^[a-z]+://(.*)$', '\\1') }}{% else %}s3-ap-south-1.amazonaws.com{% endif %}"
```
[the remark this answers is not in the document]
Comment (shobhit-vashistha) [reply ansible/roles/analytics-spark-provision/defaults/main.yml:39]: default value `s3-ap-south-1.amazonaws.com` for endpoint is hardcoded in the original `jets3t.j2` template before I made changes to it, that's the reason for setting it as default. I can either remove the default all together or move the default value to `jets3t.j2` template. Let me know which one is preferable.

File: ansible/roles/druid-ingestion/defaults/main.yml (hunk 1 above)
Comment (shobhit-vashistha) [reply ansible/roles/druid-ingestion/defaults/main.yml:2]: not specifying a group on some hardened systems (NIC servers) leads to the files being owned by `druid:root` which causes permission issue on those systems

File: ansible/roles/secor-telemetry-backup-deploy/templates/log4j.j2 (hunk 2 above)
Comment (shobhit-vashistha) [reply ansible/roles/secor-telemetry-backup-deploy/templates/log4j.j2:14]: `{{ secor_service_name }}` renders out an entire dictionary instead of the service name (same as the directory name). `{{ item.0 }}` renders the correct intended value

File: data-pipeline-flink/dp-core/src/main/scala/org/sunbird/dp/core/cache/DataCache.scala (hunk 3 above)
Comment (shobhit-vashistha) [reply data-pipeline-flink/dp-core/src/main/scala/org/sunbird/dp/core/cache/DataCache.scala:134]: `ex@(_: JedisConnectionException | _: JedisException)` in this case was present in the original commit, I had removed it to make it consistent, but I see that this change is really not necessary. Let me know, can leave this included in this PR or can leave it out.

File: data-pipeline-flink/dp-core/src/main/scala/org/sunbird/dp/core/job/BaseJobConfig.scala (hunk 4 above)
Comment (shobhit-vashistha) [reply data-pipeline-flink/dp-core/src/main/scala/org/sunbird/dp/core/job/BaseJobConfig.scala:24]: On NIC systems snappy library is not installed by the setup process, since it has to access some external url to do this access to which is restricted on their systems. Flink on the other hand runs in Kubernetes pods, which have no such issues with the snappy lib. This results in flink jobs producing messages that are compressed using snappy, but other systems running secor processes and druid ingestion, fail to read these messages because of snappy lib not being present.

Setting compression to none is a quick workaround to this issue. This change keeps the default compression as snappy, but provides an option to configure producer compression as none (or any other option).

Let me know if this is not a desirable change, will remove if that is the case, but we'll need to figure out a way to fix the snappy lib issue on NIC servers.

File: data-pipeline-flink/dp-core/src/main/scala/org/sunbird/dp/core/job/BaseJobConfig.scala (hunk 5 above)
Comment (shobhit-vashistha) [reply data-pipeline-flink/dp-core/src/main/scala/org/sunbird/dp/core/job/BaseJobConfig.scala:63]: Got it, will remove the if statements

File: ansible/inventory/env/group_vars/all.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (shobhit-vashistha) [reply ansible/inventory/env/group_vars/all.yml:29]: original file has mentioned different clusters for `ingestion_zookeeper_ip` and `processing_zookeeper_ip`

```ingestion_zookeeper_ip: "{{ groups['ingestion-cluster-zookeeper'][0] }}"
processing_zookeeper_ip: "{{ groups['processing-cluster-zookeepers'][0] }}"```

Any specific reason why same pattern should not be followed for `ingestion_kafka_brokers` also?

[the remark this answers is not in the document]
Comment (shobhit-vashistha) [reply ansible/inventory/env/group_vars/all.yml]: this change is removed from the PR

File: ansible/inventory/env/group_vars/all.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (shobhit-vashistha) [reply ansible/inventory/env/group_vars/all.yml]: this change is removed from the PR, data-pipeline only requires one

File: ansible/roles/analytics-druid/defaults/main.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (shobhit-vashistha) [reply ansible/roles/analytics-druid/defaults/main.yml]: done

File: ansible/roles/analytics-druid/defaults/main.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (shobhit-vashistha) [reply ansible/roles/analytics-druid/defaults/main.yml]: done

File: ansible/roles/analytics-spark-provision/defaults/main.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (shobhit-vashistha) [reply ansible/roles/analytics-spark-provision/defaults/main.yml]: updated and default value `s3-ap-south-1.amazonaws.com` moved to jets3t.j2 template where it was present originally

File: ansible/roles/analytics-spark-provision/templates/jets3t.j2 (hunk 11 above)
[the remark this answers is not in the document]
Comment (shobhit-vashistha) [reply ansible/roles/analytics-spark-provision/templates/jets3t.j2]: removed jinja if directive

File: ansible/roles/druid-ingestion/defaults/main.yml (hunk 1 above)
Comment (shobhit-vashistha) [reply ansible/roles/druid-ingestion/defaults/main.yml]: removed, not needed

File: data-pipeline-flink/dp-core/src/main/scala/org/sunbird/dp/core/cache/DataCache.scala (hunk 3 above)
Comment (shobhit-vashistha) [reply data-pipeline-flink/dp-core/src/main/scala/org/sunbird/dp/core/cache/DataCache.scala]: modification removed from the PR

File: data-pipeline-flink/dp-core/src/main/scala/org/sunbird/dp/core/job/BaseJobConfig.scala (hunk 5 above)
Comment (shobhit-vashistha) [reply data-pipeline-flink/dp-core/src/main/scala/org/sunbird/dp/core/job/BaseJobConfig.scala]: if statements removed

File: ansible/inventory/env/group_vars/all.yml (hunk 6 above)
Comment (shobhit-vashistha) [reply ansible/inventory/env/group_vars/all.yml]: set to private variables that will be picked up from the private repo

File: data-pipeline-flink/dp-core/src/main/scala/org/sunbird/dp/core/job/BaseJobConfig.scala (hunk 4 above)
Comment (anandp504) [reply data-pipeline-flink/dp-core/src/main/scala/org/sunbird/dp/core/job/BaseJobConfig.scala]: I understand that the compression type should be configurable. However, we can provide the default value for the configurartion as `snappy`. This allows that the projects which have not defined this configuration to `snappy` and will allow NIC to configure the compression type to `None`.

Comment (shobhit-vashistha) [reply data-pipeline-flink/dp-core/src/main/scala/org/sunbird/dp/core/job/BaseJobConfig.scala]: I have added snappy as the default, the code snippet above is outdated

**Review by anandp504 — APPROVED** [review]
Approved
